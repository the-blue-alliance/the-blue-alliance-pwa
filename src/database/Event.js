import { Record } from "immutable";
// import moment from 'moment-timezone'

export const REGIONAL = 0;
export const DISTRICT = 1;
export const DISTRICT_CMP = 2;
export const CMP_DIVISION = 3;
export const CMP_FINALS = 4;
export const DISTRICT_CMP_DIVISION = 5;
export const FOC = 6;
export const OFFSEASON = 99;
export const PRESEASON = 100;

// const CMP_TYPES = new Set([
//   CMP_DIVISION,
//   CMP_FINALS,
// ])

// const OFFICIAL_TYPES = new Set([
//   REGIONAL,
//   DISTRICT,
//   DISTRICT_CMP_DIVISION,
//   DISTRICT_CMP,
//   CMP_DIVISION,
//   CMP_FINALS,
//   FOC,
// ])

export default class Event extends Record({
  key: undefined,
  name: undefined,
  short_name: undefined,
  year: undefined,
  event_code: undefined,
  week: undefined,
  event_type: undefined,
  district: undefined,
  start_date: undefined,
  end_date: undefined,
  timezone: undefined,
  address: undefined,
  city: undefined,
  state_prov: undefined,
  country: undefined,
  location_name: undefined,
  gmaps_url: undefined,
  website: undefined,
  webcasts: undefined,
  playoff_type: undefined,
}) {
  // Name
  safeShortName() {
    return this.short_name ? this.short_name : this.name;
  }

  // Location
  getCityStateCountry() {
    if (this.cityStateCountry === undefined) {
      this.cityStateCountry = "";
      if (this.city) {
        this.cityStateCountry += `${this.city}`;
      }
      if (this.state_prov) {
        this.cityStateCountry += `, ${this.state_prov}`;
      }
      if (this.country) {
        this.cityStateCountry += `, ${this.country}`;
      }
      if (this.cityStateCountry === "") {
        this.cityStateCountry = null;
      }
    }
    return this.cityStateCountry;
  }

  // getCityStateCountryLower() {
  //   if (this.cityStateCountryLower === undefined) {
  //     if (this.getCityStateCountry()) {
  //       this.cityStateCountryLower = this.getCityStateCountry().toLowerCase()
  //     } else {
  //       this.cityStateCountryLower = null
  //     }
  //   }
  //   return this.cityStateCountryLower
  // }

  // // Time
  // getDateString() {
  //   if (this.dateStr === undefined) {
  //     const startDate = moment(this.start_date)
  //     const endDate = moment(this.end_date)
  //     this.dateStr = endDate.format('MMM D, YYYY').replace(/ /g, '\u00a0')
  //     if (this.start_date !== this.end_date) {
  //       const startDateStr = startDate.format('MMM D')
  //       this.dateStr = `${startDateStr.replace(/ /g, '\u00a0')} to ${this.dateStr.replace(/ /g, '\u00a0')}`
  //     }
  //   }
  //   return this.dateStr
  // }
  //
  // startMoment() {
  //   return moment.tz(this.start_date, this.timezone)
  // }
  //
  // endMoment() {
  //   // Add one day because end_date is 12 AM
  //   return moment.tz(this.end_date, this.timezone).add(1, 'days')
  // }

  // withinDays(negativeDaysBefore, daysAfter) {
  //   const now = moment.now()
  //   const afterStart = this.startMoment().add(negativeDaysBefore, 'days') < now
  //   const beforeEnd = this.endMoment().add(daysAfter, 'days') > now
  //   return afterStart && beforeEnd
  // }
  //
  // isNow() {
  //   return !this.isPast() && !this.isFuture()
  // }
  //
  // isPast() {
  //   return this.endMoment() < moment.now()
  // }
  //
  // isFuture() {
  //   return this.startMoment() > moment.now()
  // }

  // isThisWeek() {
  //   // An event is this week iff
  //   // 1) The event is within 1 day of now
  //   // OR
  //   // 2) The event start date is on or within 4 days of the closest Wednesday
  //   if (this.withinDays(-1, 1)) {
  //     return true
  //   }
  //   const wed = 4
  //   const thisWed = moment().isoWeekday(4)
  //   const nextWed = moment().add(1, 'weeks').isoWeekday(wed)
  //   const today = moment.now()
  //   let closestWed
  //   if (Math.abs(thisWed.diff(today)) < Math.abs(nextWed.diff(today))) {
  //     closestWed = thisWed
  //   } else {
  //     closestWed = nextWed
  //   }
  //   const offset = moment.duration(this.startMoment() - closestWed).asDays()
  //   return Math.abs(offset) <= 4
  // }

  // // Event type
  // isCMP() {
  //   return CMP_TYPES.has(this.event_type)
  // }

  // isDistrictQual() {
  //   return this.event_type === DISTRICT
  // }

  // isFOC() {
  //   return this.event_type === FOC
  // }

  // isPreseason() {
  //   return this.event_type === PRESEASON
  // }

  // isOfficial() {
  //   return OFFICIAL_TYPES.has(this.event_type)
  // }

  isRegional() {
    return this.event_type === REGIONAL;
  }
}
