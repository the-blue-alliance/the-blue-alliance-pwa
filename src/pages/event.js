import React from "react";
import PropTypes from "prop-types";
import { getEventFetchStatus, getEvent } from "../selectors/EventSelectors";
import {
  getEventMatchesFetchStatus,
  getEventMatches,
} from "../selectors/MatchSelectors";
import { fetchEvent, fetchEventMatches } from "../actions";
import useData from "../lib/useData";
import notFoundError from "../lib/notFoundError";
import Typography from "@material-ui/core/Typography";
import Page from "../components/Page";
import MatchRow from "../components/MatchRow";
import EventPageDialog from "../components/EventPageDialog";

const Event = ({ eventKey, refetchOnLoad }) => {
  const [event, eventFetchStatus, refetchEvent] = useData(
    state => getEventFetchStatus(state, eventKey),
    state => getEvent(state, eventKey),
    React.useMemo(() => fetchEvent(eventKey), [eventKey]),
    refetchOnLoad.event
  );
  const [unsortedMatches, matchesFetchStatus, refetchMatches] = useData(
    state => getEventMatchesFetchStatus(state, eventKey),
    state => getEventMatches(state, eventKey),
    React.useMemo(() => fetchEventMatches(eventKey), [eventKey]),
    refetchOnLoad.eventMatches
  );
  const handleRefresh = React.useCallback(() => {
    refetchEvent();
    refetchMatches();
  }, [refetchEvent, refetchMatches]);

  // Sort matches
  const matches = React.useMemo(
    () =>
      unsortedMatches.sort((a, b) => {
        return a.getNaturalOrder() - b.getNaturalOrder();
      }),
    [unsortedMatches]
  );

  if (!event) {
    return notFoundError();
  }

  return (
    <Page
      title={`${event.name} (${event.year})`}
      metaDescription={`Videos and match results for the ${event.year} ${
        event.name
      } FIRST Robotics Competition in ${event.getCityStateCountry()}.`}
      additionalMetas={
        <>
          <meta property="og:location" content={event.getCityStateCountry()} />
          <meta property="og:start_time" content={event.start_date} />
          <meta property="og:end_time" content={event.end_date} />
        </>
      }
      isLoading={
        eventFetchStatus === "fetching" || matchesFetchStatus === "fetching"
      }
      refreshFunction={handleRefresh}
    >
      <Typography variant="h4">{event.name}</Typography>
      {matches.map(match => (
        <MatchRow key={match.key} eventKey={eventKey} match={match} />
      ))}
      <EventPageDialog eventKey={eventKey} />
    </Page>
  );
};

Event.getInitialProps = async ({ reduxStore, query }) => {
  const eventKey = query.eventKey;
  const state = reduxStore.getState();

  const eventFetchInitial = getEventFetchStatus(state, eventKey) !== "success";
  const eventMatchesFetchInitial =
    getEventMatchesFetchStatus(state, eventKey) !== "success";

  const fetchPromises = [];
  if (eventFetchInitial) {
    fetchPromises.push(reduxStore.dispatch(fetchEvent(eventKey)));
  }
  if (eventMatchesFetchInitial) {
    fetchPromises.push(reduxStore.dispatch(fetchEventMatches(eventKey)));
  }
  await Promise.all(fetchPromises);

  return {
    eventKey,
    refetchOnLoad: {
      event: !eventFetchInitial,
      eventMatches: !eventMatchesFetchInitial,
    },
  };
};

Event.propTypes = {
  eventKey: PropTypes.string,
  refetchOnLoad: PropTypes.object,
};

export default Event;
