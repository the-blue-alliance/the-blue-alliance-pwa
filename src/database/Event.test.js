/* eslint-env jest */
import { fromJS } from "immutable";
import Event from "./Event";
import { newEvent } from "../reducers/tests/data/event"; // TODO: Don't use test data from here

describe("Event record", () => {
  describe("typical model", () => {
    const event = new Event(fromJS(newEvent));

    it("has correct safeShortName", () => {
      expect(event.safeShortName()).toEqual("Silicon Valley");
    });

    it("has correct cityStateCountry", () => {
      expect(event.getCityStateCountry()).toEqual("San Jose, CA, USA");
      // Check for proper caching
      expect(event.cityStateCountry).toEqual("San Jose, CA, USA");
      expect(event.getCityStateCountry()).toEqual("San Jose, CA, USA");
    });

    it("has correct isRegional", () => {
      expect(event.isRegional()).toBe(true);
    });

    it("has expected structured data", () => {
      expect(event.structuredData()).toMatchSnapshot();
    });
  });

  describe("null model", () => {
    // Tests minimally populated Event
    const event = new Event(
      fromJS({
        name: "Silicon Valley Regional",
      })
    );

    it("has correct safeShortName", () => {
      expect(event.safeShortName()).toEqual("Silicon Valley Regional");
    });

    it("has correct cityStateCountry", () => {
      expect(event.getCityStateCountry()).toBe(null);
      // Check for proper caching
      expect(event.cityStateCountry).toEqual(null);
      expect(event.getCityStateCountry()).toEqual(null);
    });
  });
});
