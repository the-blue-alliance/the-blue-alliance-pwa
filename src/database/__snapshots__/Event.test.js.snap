// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event record typical model has expected structured data 1`] = `
<script
  dangerouslySetInnerHTML={
    Object {
      "__html": "{\\"@context\\":\\"http://schema.org\\",\\"@type\\":\\"Event\\",\\"name\\":\\"Silicon Valley Regional\\",\\"startDate\\":\\"2019-03-28T00:00:00.000-07:00\\",\\"endDate\\":\\"2019-03-31T23:59:59.000-07:00\\",\\"location\\":{\\"@type\\":\\"Place\\",\\"name\\":\\"The Event Center at SJSU\\",\\"address\\":{\\"@type\\":\\"PostalAddress\\",\\"streetAddress\\":\\"290 S 7th St, San Jose, CA 95112, USA\\",\\"addressLocality\\":\\"San Jose\\",\\"addressRegion\\":\\"CA\\",\\"addressCountry\\":\\"USA\\",\\"postalCode\\":\\"95112\\"}}}",
    }
  }
  type="application/ld+json"
/>
`;
